Implement SongTicker component in main menu

EVAST9919 opened the pull request:
Fixes https://github.com/ppy/osu/issues/7274

![Screenshot_3](https://user-images.githubusercontent.com/22874522/72198743-a80afe00-3442-11ea-904e-29abebb387dd.png)

EVAST9919 committed:
Implement SongTicker component

EVAST9919 committed:
Move only ButtonSystem on screen changes rather than everything

EVAST9919 committed:
Some animation adjustments

EVAST9919 committed:
Add unicode metadata support

peppy commented on osu.Game/Screens/Menu/MainMenu.cs:
```diff
@@ -226,15 +236,17 @@ public override void OnSuspending(IScreen next)
         public override void OnResuming(IScreen last)
         {
             base.OnResuming(last);
+            songTicker.Hide();
```
this shouldn't be necessary

EVAST9919 replied:
song ticker can be triggered even in song select screen. So if you are in there, select another beatmap and go back to main menu - you will see the song ticker. That's why I'm hiding it here

peppy replied:
it shouldn't be processing changes when not at the menu.

peppy replied:
and how does it even become visible again then, after returning from song select?

EVAST9919 replied:
Good question. Exiting song select without that line:
![nice](https://user-images.githubusercontent.com/22874522/72206179-d5d06100-349b-11ea-96a7-4bf3e1f6616c.gif)

peppy replied:
As long as you understand it's a question you need to answer. The `Hide` as it stands is not correct.

EVAST9919 replied:
Got it, investigating

EVAST9919 replied:
Simplest (but not sure if correct) solution: I just added `AllowUpdates` flag to song ticker.

iiSaLMaN replied:
You're probably looking for `mainMenu.IsCurrentScreen()`, or even more simpler, checking if the ticker `IsPresent`.

EVAST9919 replied:
`IsPresent` will not work because even on the main screen we are triggering changes while fully invisible

iiSaLMaN replied:
Right, in that case, pass the menu and check `IsCurrentScreen()`.

EVAST9919 replied:
I personally don't think passing a screen for the component is a good idea. What if (for example) it will be implemented somewhere else? There may be other conditions.

iiSaLMaN replied:
That'll be added later, currently it is used once at the main menu and it should not update when it's out of the screen, no need to make it that complex in my opinion.

peppy commented on osu.Game/Screens/Menu/MainMenu.cs:
```diff
@@ -263,6 +275,7 @@ public override bool OnExiting(IScreen next)
 
             buttons.State = ButtonSystemState.Exit;
             this.FadeOut(3000);
+            songTicker.Hide();
```
nor should this

EVAST9919 replied:
And here It can be visible too

EVAST9919 replied:
That's why this exactly one is required
https://streamable.com/jecpp

peppy commented on osu.Game/Screens/Menu/SongTicker.cs:
```diff
@@ -0,0 +1,72 @@
+        [Resolved]
+        private Bindable<WorkingBeatmap> beatmap { get; set; }
+
+        private readonly Bindable<WorkingBeatmap> workingBeatmap = new Bindable<WorkingBeatmap>();
```
this is not how bindables work. you do not need this field.

iiSaLMaN replied:
is this due to `SongTicker` never getting removed?

EVAST9919 replied:
In one pr someone told me to use local bindable rather than using `BindValueChanged` for resolved one.

peppy replied:
https://github.com/ppy/osu-framework/blob/master/osu.Framework/Allocation/ResolvedAttribute.cs#L115

peppy commented on osu.Game/Screens/Menu/SongTicker.cs:
```diff
@@ -0,0 +1,72 @@
+
+        private void onBeatmapChanged(ValueChangedEvent<WorkingBeatmap> working)
+        {
+            if (working.NewValue?.Beatmap == null)
```
workingbeatmap can't ever be null (check game/gamebase implementation)

EVAST9919 replied:
saw that in the MusicController. May be I misunderstood something

peppy replied:
it's unnecessary there too. see https://github.com/ppy/osu/blob/master/osu.Game/OsuGameBase.cs#L208

peppy commented on osu.Game/Screens/Menu/SongTicker.cs:
```diff
@@ -0,0 +1,72 @@
+            if (working.NewValue?.Beatmap == null)
+                return;
+
+            var metadata = working.NewValue?.Metadata;
```
remove null coalesce. it's not valid as above, and also having it present makes the next two lines 100% incorrect.

iiSaLMaN commented on osu.Game/Screens/Menu/MainMenu.cs:
```diff
@@ -69,12 +69,22 @@ public class MainMenu : OsuScreen
             holdDelay = config.GetBindable<float>(OsuSetting.UIHoldActivationDelay);
             loginDisplayed = statics.GetBindable<bool>(Static.LoginOverlayDisplayed);
 
+            AddInternal(songTicker = new SongTicker
```
I don't like how the children are added in splitted way, makes the code unnecessarily long, mabye add this to the `AddRangeInternal` below.

EVAST9919 replied:
No. It should be below `ExitConfirmOverlay`

EVAST9919 committed:
Merge remote-tracking branch 'refs/remotes/ppy/master' into main-menu-song-ticker

EVAST9919 committed:
Remove unnecessary local bindable

EVAST9919 committed:
remove unnecessary null checks

EVAST9919 committed:
Add AllowUpdates flag to SongTicker

EVAST9919 committed:
Merge remote-tracking branch 'refs/remotes/ppy/master' into main-menu-song-ticker

EVAST9919 committed:
Fade out instead of insta hiding on menu suspending

peppy committed:
Merge remote-tracking branch 'upstream/master' into main-menu-song-ticker

peppy committed:
Tidy up implementation and show on returning with new song

peppy committed:
Ensure only run once when not current screen

peppy committed:
Add a test scene

You have to have a test scene You have to have a test scene You have to have a test scene You have to have a test scene You have to have a test scene You have to have a test scene You have to have a test scene You have to have a test scene You have to have a test scene You have to have a test scene You have to have a test scene You have to have a test scene You have to have a test scene

peppy committed:
Adjust fade

Joehuu commented on osu.Game/Screens/Menu/MainMenu.cs:
```diff
@@ -107,8 +96,28 @@ private void load(DirectOverlay direct, SettingsOverlay settings, OsuConfigManag
+                    }
+                });
+            }
+
```
You broke `ButtonSystem`'s `OnPressed` when moving this.

peppy replied:
aaaaaaaa i forgot to push 
![image](https://user-images.githubusercontent.com/191335/73054158-061cf400-3ecd-11ea-9eb3-5149c795bfe7.png)
